lib/ansible/modules/cloud/docker/docker_stack.py

```diff
@@ -0,0 +1,226 @@
+            This will have docker remove the services not present in the
+            current stack definition.
+
+requirements:
+-   "jsondiff"
+'''
+
+RETURN = '''
+docker_stack_spec_diff:
+    description: |
+        dictionary containing the differences between the 'Spec' field
+        of the stack services before and after applying the new stack
+        definition
+    sample: >
+        "docker_stack_specs_diff":
+        {'test_stack_test_service': {u'TaskTemplate': {u'ContainerSpec': {delete: [u'Env']}}}}
+    returned: on change
+    type: dict
+'''
+
+EXAMPLES = '''
+-   name: deploy 'stack1' stack from file
+    docker_stack:
+        state: present
+        name: stack1
+        compose_file: /opt/stack.compose
+-   name: deploy 'stack2' from yaml
+    docker_stack:
+        state: present
+        name: stack2
+        compose_yaml: |
+            version: '3'
+            services:
+                web:
+                    image: nginx
+                    ports:
+                    -   "80:80"
+-   name: deprovision 'stack1'
+    docker_stack:
+        state: absent
+'''
+
+from ansible.module_utils.basic import AnsibleModule, os
+from tempfile import mkstemp
+import json
+from jsondiff import diff as json_diff
+
+
+def docker_stack_services(module, stack_name):
+    rc, out, err = module.run_command(["docker",
+                                       "stack",
+                                       "services",
+                                       stack_name,
+                                       "--format",
+                                       "{{.Name}}"])
+    if out != ("Nothing found in stack %s\n" % stack_name):
+        return(out.strip().split('\n'))
+    return []
+
+
+def docker_service_inspect(module, service_name):
+    rc, out, err = module.run_command(["docker",
+                                       "service",
+                                       "inspect",
+                                       service_name])
+    if rc != 0:
+        return None
+    else:
+        ret = json.loads(out)[0]['Spec']
+        return(ret)
+
+
+def docker_stack_deploy(module, stack_name, compose_file):
+    command = ["docker", "stack", "deploy"]
+    if module.params["prune"]:
+        command += ["--prune"]
+    command += ["--compose-file",
+                compose_file,
+                stack_name]
+    return module.run_command(command)
+
+
+def docker_stack_inspect(module, stack_name):
+    ret = {}
+    for service_name in docker_stack_services(module, stack_name):
+        ret[service_name] = docker_service_inspect(module, service_name)
+    return ret
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec={
+            'name': dict(required=True),
+            'compose_yaml': dict(),
+            'compose_file': dict(),
+            'prune': dict(default=False, type='bool'),
+            'state': dict(default='present', choices=['present', 'absent'])
```

Comment:
state in `DOCUMENTATION` say required but I see this is not `required=True` here.